openedx/features/qverse_features/registration/signals.py

```diff
@@ -1,20 +1,182 @@
 
-UserInfo = namedtuple('UserInfo',
-                      ['username', 'email','first_name', 'last_name',
-                       'other_name', 'mobile_number', 'level',
-                       'programme_and_department']
-                     )
+
+LOGGER = logging.getLogger(__name__)
 
 
 @receiver(post_save, sender=BulkUserRegistration)
-def make_users_from_csv_file(sender, instance, created, **kwargs):
-    # TODO: this would be implemented later once the basic architecture is final.
-    pass
+def create_users_from_csv_file(sender, instance, created, **kwargs):
+    """
+    Creates/Updates users and their profiles.
+
+    Reads data from given csv file having user details. Then creates/updates
+    edx user, edx user profile and qverse user profile accordingly. It also
+    writes the status of creation/update on the given file by adding an
+    extra column named 'status' in it. Also handles the exceptions raised by
+    the functions that it calls.
+
+    Arguments:
+        sender (ModelBase): responsible to initiate the signal
+        instance (BulkUserRegistration): newly created/updated object of BulkUserRegistration
+        created (bool): Is created/updated?
+        kwargs (dict): Other info
+    """
+    csv_file = None
+    try:
+        csv_file = open(instance.admission_file.path, 'r')
+    except IOError as error:
+        LOGGER.exception('({}) --- {}'.format(error.filename, error.strerror))
+        return
+    dict_reader = DictReader(csv_file, delimiter=',')
+    # The header of uploaded csv might be in uppercase, lowercase or in camel case
+    # so to make our code independent to case just turn all the keys of reader into lower case.
+    # Also in python standard library for csv there is an option named skipinitialwhitespaces
+    # which is only applied for leading whitespaces and there is no option for trailing
+    # whitespaces. So, we will have to handle that case ourselves
+    reader = (dict((k.strip().lower(), v.strip() if v else v) for k, v in row.items()) for row in dict_reader)
+    output_file_rows = []
+    try:
+        for row in reader:
+            try:
+                output_file_rows.append(row)
+                output_file_rows[-1].update({'status': ''})
+                _validate_single_csv_row(row)
+                # Creating/Updating edX User, edX User Profile and QVerse User Profile
+                # will be an atomic operation. If one operation fails the previous
+                # successfull operations will be reverted and no changes will be applied.
+                with transaction.atomic():
+                    edx_user = _create_or_update_edx_user(row)
+                    _create_or_update_edx_user_profile(edx_user)
+                    _create_or_update_qverse_user_profile(edx_user, row, output_file_rows)
+            except (IntegrityError, ValueError):
+                LOGGER.exception('Error while creating/updating ({}) user and its profiles.'.format(row.get('regno')))
```

Comment:
this error should be explicit.

Reply:
errors should be handled in respective methods.

Reply:
LOGGER.exception will handle most of the job, but still I have added extra description for the errors.

Reply:
Logging message printing by my logger statement
![image](https://user-images.githubusercontent.com/42294172/77429872-23ead480-6dfc-11ea-8819-4bf73cc4f4bb.png)

Reply:
LOGGER.exception also printing the same error message

![image](https://user-images.githubusercontent.com/42294172/77430084-70ceab00-6dfc-11ea-9440-6aa000fd5216.png)

Reply:
@tasawernawaz these are the screenshots which you said.

but these are only for failure scenario but now you are asking for success scenerio too so now I'm uploading it

Reply:
Success scenerio

![image](https://user-images.githubusercontent.com/42294172/77509435-842a5680-6e8e-11ea-9ae1-bc8421a50c8d.png)